File: sdk-node/src/redact.ts
/**
 * Local PII redaction — runs inside the user's process BEFORE anything leaves
 * the app. Pantau's servers never see raw bodies/headers.
 *
 * Two tiers:
 *  - MASK  → value replaced with '[REDACTED]'. Not searchable, not readable.
 *            For secrets (password, token, card) — no reason to ever query them.
 *  - HASH  → value replaced with '#<hmac16>'. Not readable, but SEARCHABLE:
 *            the same input always yields the same hash, so you can look up
 *            "did user X hit /login" by hashing X the same way. HMAC is keyed
 *            with the account api_key so a DB leak can't be reversed.
 */
import { createHmac } from 'crypto';

const DENY_HEADERS = new Set([
  'authorization', 'cookie', 'set-cookie', 'x-api-key', 'proxy-authorization',
]);

// Full-mask keys: secrets, never searchable.
const DEFAULT_DENY_KEYS = [
  'password', 'passwd', 'secret', 'token', 'apikey', 'api_key',
  'authorization', 'cookie', 'credit_card', 'card_number', 'cardnumber',
  'cvv', 'cvc', 'ssn', 'pin', 'private_key', 'access_token', 'refresh_token',
];

// Hash keys: PII you may need to search on (GDPR / UU PDP) — stored as a
// keyed hash, searchable but not readable.
const DEFAULT_HASH_KEYS = ['email', 'phone', 'nik', 'passport'];

// Partial keys: PII shown partly (b***@gmail.com). Readable-ish, NOT reliably
// searchable. Empty by default — opt in via capture.partialKeys.
const DEFAULT_PARTIAL_KEYS: string[] = [];

export const REDACTED = '[REDACTED]';
const MAX_DEPTH = 8;

export interface RedactOptions {
  denyKeys?: string[];
  hashKeys?: string[];
  partialKeys?: string[];
  hashSecret?: string; // HMAC key (the account api_key). Without it, hash keys fall back to mask.
}

function matches(key: string, list: string[]): boolean {
  const k = key.toLowerCase();
  return list.some((d) => k.includes(d));
}

/** Deterministic keyed hash of a value: '#' + first 16 hex of HMAC-SHA256. */
export function hashValue(value: unknown, secret: string): string {
  const norm = String(value).trim().toLowerCase();
  return '#' + createHmac('sha256', secret).update(norm).digest('hex').slice(0, 16);
}

/**
 * Partially mask a value, keeping just enough to eyeball it:
 *  - email:  b***@gmail.com   (first char + domain)
 *  - other:  keep last 2 chars, e.g. ****89
 */
export function partialMask(value: unknown): string {
  const s = String(value);
  const at = s.indexOf('@');
  if (at > 0) {
    const local = s.slice(0, at);
    const domain = s.slice(at);
    const head = local[0];
    return `${head}***${domain}`;
  }
  if (s.length <= 2) return '***';
  return '***' + s.slice(-2);
}

/** Deep-clone `value`, masking secrets and hashing searchable PII. */
export function redactBody(value: unknown, opts: RedactOptions = {}, depth = 0): unknown {
  const denyKeys = opts.denyKeys ?? DEFAULT_DENY_KEYS;
  const hashKeys = opts.hashKeys ?? DEFAULT_HASH_KEYS;
  const partialKeys = opts.partialKeys ?? DEFAULT_PARTIAL_KEYS;

  if (depth > MAX_DEPTH || value == null) return value;

  if (Array.isArray(value)) {
    return value.map((v) => redactBody(v, opts, depth + 1));
  }

  if (typeof value === 'object') {
    const out: Record<string, unknown> = {};
    for (const [k, v] of Object.entries(value as Record<string, unknown>)) {
      const isPrimitive = v != null && typeof v !== 'object';
      if (matches(k, denyKeys)) {
        out[k] = REDACTED;
      } else if (matches(k, partialKeys)) {
        // Partial takes priority over hash: readable-ish beats searchable.
        out[k] = isPrimitive ? partialMask(v) : redactBody(v, opts, depth + 1);
      } else if (matches(k, hashKeys)) {
        // Only hash primitives; hash secret required, else mask.
        out[k] = opts.hashSecret != null && isPrimitive
          ? hashValue(v, opts.hashSecret)
          : (isPrimitive ? REDACTED : redactBody(v, opts, depth + 1));
      } else {
        out[k] = redactBody(v, opts, depth + 1);
      }
    }
    return out;
  }

  return value;
}

export function redactHeaders(headers: Record<string, unknown>): Record<string, unknown> {
  const out: Record<string, unknown> = {};
  for (const [k, v] of Object.entries(headers || {})) {
    out[k] = DENY_HEADERS.has(k.toLowerCase()) ? REDACTED : v;
  }
  return out;
}

File: sdk-python/pantau/redact.py
"""Local PII redaction — runs inside user's process before data leaves the app.

Two tiers:
  - MASK  → value replaced with '[REDACTED]'. Not searchable.
  - HASH  → '#<hmac16>'. Searchable, keyed with account api_key.
"""

import hashlib
import hmac
from typing import Any, Dict, List, Optional

REDACTED = "[REDACTED]"

DENY_HEADERS = {
    "authorization", "cookie", "set-cookie", "x-api-key",
    "proxy-authorization",
}

DEFAULT_DENY_KEYS = [
    "password", "passwd", "secret", "token", "apikey", "api_key",
    "authorization", "cookie", "credit_card", "card_number", "cardnumber",
    "cvv", "cvc", "ssn", "pin", "private_key", "access_token", "refresh_token",
]

DEFAULT_HASH_KEYS = ["email", "phone", "nik", "passport"]

DEFAULT_PARTIAL_KEYS: List[str] = []

MAX_DEPTH = 8


def _matches(key: str, patterns: List[str]) -> bool:
    k = key.lower()
    return any(p in k for p in patterns)


def hash_value(value: Any, secret: str) -> str:
    """Deterministic keyed hash: '#' + first 16 hex of HMAC-SHA256."""
    norm = str(value).strip().lower()
    h = hmac.new(secret.encode("utf-8"), norm.encode("utf-8"), hashlib.sha256)
    return "#" + h.hexdigest()[:16]


def partial_mask(value: Any) -> str:
    """Partially mask: b***@gmail.com or ****89."""
    s = str(value)
    at = s.find("@")
    if at > 0:
        local = s[:at]
        domain = s[at:]
        head = local[0] if local else "x"
        return f"{head}***{domain}"
    if len(s) <= 2:
        return "***"
    return "***" + s[-2:]


def redact_body(
    value: Any,
    opts: Optional[Dict] = None,
    depth: int = 0,
) -> Any:
    """Deep-clone `value`, masking secrets and hashing searchable PII."""
    if opts is None:
        opts = {}

    deny_keys = opts.get("denyKeys") or DEFAULT_DENY_KEYS
    hash_keys = opts.get("hashKeys") or DEFAULT_HASH_KEYS
    partial_keys = opts.get("partialKeys") or DEFAULT_PARTIAL_KEYS
    hash_secret = opts.get("hashSecret")

    if depth > MAX_DEPTH or value is None:
        return value

    if isinstance(value, list):
        return [redact_body(v, opts, depth + 1) for v in value]

    if isinstance(value, dict):
        out = {}
        for k, v in value.items():
            is_primitive = isinstance(v, (str, int, float, bool, type(None)))
            if _matches(k, deny_keys):
                out[k] = REDACTED
            elif _matches(k, partial_keys):
                out[k] = partial_mask(v) if is_primitive else redact_body(v, opts, depth + 1)
            elif _matches(k, hash_keys):
                if hash_secret and is_primitive:
                    out[k] = hash_value(v, hash_secret)
                elif is_primitive:
                    out[k] = REDACTED
                else:
                    out[k] = redact_body(v, opts, depth + 1)
            else:
                out[k] = redact_body(v, opts, depth + 1)
        return out

    return value


def redact_headers(headers: Dict[str, Any]) -> Dict[str, Any]:
    """Redact sensitive headers in-place."""
    if not headers:
        return {}
    return {
        k: REDACTED if k.lower() in DENY_HEADERS else v
        for k, v in headers.items()
    }
